drivers/dma/dma_silabs_siwx917.c

```diff
@@ -0,0 +1,480 @@
+	RSI_UDMA_HANDLE_T udma_handle = NULL;
+	/* Initialize UDMA Resources */
+	UDMA_RESOURCES UDMA_Siwx917_Resources = {
+		.reg = cfg->reg, /* UDMA register base address */
+		.udma_irq_num = cfg->irq_number,
+	};
+
+	if (cfg->reg == UDMA0) {
+		UDMA_Siwx917_Resources.desc =
+			(RSI_UDMA_DESC_T *)DT_PROP(DT_NODELABEL(udma0), sram_desc_addr);
+	} else if (cfg->reg == UDMA1) {
+		UDMA_Siwx917_Resources.desc =
+			(RSI_UDMA_DESC_T *)DT_PROP(DT_NODELABEL(udma1), sram_desc_addr);
+	} else {
+		return -EINVAL;
+	}
+
+	udma_handle = UDMAx_Initialize(&UDMA_Siwx917_Resources, UDMA_Siwx917_Resources.desc,
+				       udma_handle, data->dma_rom_buff);
+	if (udma_handle != (RSI_UDMA_HANDLE_T)data->dma_rom_buff) {
+		return -EINVAL;
+	}
+
+	/* Connect the DMA interrupt */
+	cfg->irq_configure();
+
+	/* Enable UDMA instance */
+	if (UDMAx_DMAEnable(&UDMA_Siwx917_Resources, udma_handle) != 0) {
+		return -EBUSY;
+	}
+	return 0;
+}
+
+static void dma_siwx917_isr(const struct device *dev)
+{
+	const struct dma_siwx917_config *cfg = dev->config;
+	struct dma_siwx917_data *data = dev->data;
+	uint32_t irq_number = cfg->irq_number;
+	uint8_t transfer_complete = 0;
+	uint8_t soft_trig_flag = 0;
+	uint32_t int_status;
+	uint8_t channel;
+	/* Initialize UDMA Resources */
+	UDMA_RESOURCES UDMA_Siwx917_Resources = {
+		cfg->reg, irq_number,
+		(RSI_UDMA_DESC_T *)cfg->reg->CTRL_BASE_PTR /* SRAM base address */
+	};
+
+	/* Disable the IRQ to prevent the ISR from being triggered by */
+	/* interrupts from other DMA channels */
+	irq_disable(irq_number);
+	int_status = cfg->reg->UDMA_DONE_STATUS_REG; /* Read the interrupt status */
+	/* Identify the interrupt channel */
+	for (channel = 0; channel < cfg->channels; channel++) {
+		if (!(int_status & (1U << channel))) {
+			continue;
+		}
+		if (data->udma_channel_info[channel].Cnt == data->udma_channel_info[channel].Size) {
+			transfer_complete = 1;
+			break;
+		}
+		/* Check if the transfer type is memory-memory */
+		if ((UDMA_Siwx917_Resources.desc[channel].vsUDMAChaConfigData1.srcInc !=
+		     UDMA_SRC_INC_NONE) &&
+		    (UDMA_Siwx917_Resources.desc[channel].vsUDMAChaConfigData1.dstInc !=
+		     UDMA_DST_INC_NONE)) {
+			/* Need to apply a software trigger later */
+			soft_trig_flag = 1;
+		}
+		break;
+	}
+	if (transfer_complete) {
+		if (data->dma_callback != NULL) {
+			/* Transfer complete, call user callback */
+			data->dma_callback(dev, data->cb_data, channel + 1, 0);
+		}
+		cfg->reg->UDMA_DONE_STATUS_REG = (1U << channel);
+	} else {
+		/* Call UDMA ROM IRQ handler. */
+		ROMAPI_UDMA_WRAPPER_API->uDMAx_IRQHandler(&UDMA_Siwx917_Resources,
+							  UDMA_Siwx917_Resources.desc,
+							  data->udma_channel_info);
+		if (soft_trig_flag) {
+			/* Set the software trigger bit for starting next transfer */
+			cfg->reg->CHNL_SW_REQUEST |= (1U << channel);
+		}
+	}
+	/* Enable the IRQ to restore interrupt functionality for other DMA channels */
+	irq_enable(irq_number);
+}
```

Comment:
With a size use of `find_lsb_set()`, you can greatly simply this function:
```
static void dma_siwx917_isr(const struct device *dev)
{
	const struct dma_siwx917_config *cfg = dev->config;
	struct dma_siwx917_data *data = dev->data;
	UDMA_RESOURCES ures = {
		.reg = cfg->reg,
		.udma_irq_num = cfg->irq_number,
		.desc = cfg->reg->CTRL_BASE_PTR,
	};
	uint8_t channel;

	irq_disable(cfg->irq_number);
	channel = find_lsb_set(cfg->reg->UDMA_DONE_STATUS_REG);
	if (!channel || channel > cfg->channels) {
		goto out;
	}
        /* find_lsb_set() returns 1 indexed value */
	channels -= 1;
	if (data->udma_channel_info[channel].Cnt == data->udma_channel_info[channel].Size) {
		if (data->dma_callback) {
			data->dma_callback(dev, data->cb_data, channel + 1, 0);
		}
		cfg->reg->UDMA_DONE_STATUS_REG = BIT(channel);
	} else {
		ROMAPI_UDMA_WRAPPER_API->uDMAx_IRQHandler(&ures, cfg->reg->CTRL_BASE_PTR,
							  data->udma_channel_info);
		/* Is a Memory-to-memory DMA */
		if (ures.desc[channel].vsUDMAChaConfigData1.srcInc != UDMA_SRC_INC_NONE &&
		    ures.desc[channel].vsUDMAChaConfigData1.dstInc != UDMA_DST_INC_NONE) {
			cfg->reg->CHNL_SW_REQUEST |= BIT(channel);
		}
	}
out:
	irq_enable(cfg->irq_number);
}
```

Reply:
You're welcome! I'm glad the suggestion helped simplify the function